File: MusicTestProject/utNota.vb
Imports System.Text
Imports Microsoft.VisualStudio.TestTools.UnitTesting

<TestClass()>
Public Class utNota

    Private testContextInstance As TestContext

    '''<summary>
    '''Gets or sets the test context which provides
    '''information about and functionality for the current test run.
    '''</summary>
    Public Property TestContext() As TestContext
        Get
            Return testContextInstance
        End Get
        Set(ByVal value As TestContext)
            testContextInstance = Value
        End Set
    End Property

#Region "Additional test attributes"
    '
    ' You can use the following additional attributes as you write your tests:
    '
    ' Use ClassInitialize to run code before running the first test in the class
    ' <ClassInitialize()> Public Shared Sub MyClassInitialize(ByVal testContext As TestContext)
    ' End Sub
    '
    ' Use ClassCleanup to run code after all tests in a class have run
    ' <ClassCleanup()> Public Shared Sub MyClassCleanup()
    ' End Sub
    '
    ' Use TestInitialize to run code before running each test
    ' <TestInitialize()> Public Sub MyTestInitialize()
    ' End Sub
    '
    ' Use TestCleanup to run code after each test has run
    ' <TestCleanup()> Public Sub MyTestCleanup()
    ' End Sub
    '
#End Region

    <TestMethod()>
    Public Sub VerificaNoteStatiche()
        Assert.AreEqual(Nota.[DO], New Nota(NotaEnum.[DO]), "Nota statica DO non corrisponde a quella del costruttore")
        Assert.AreEqual(Nota.DOdiesis, New Nota(NotaEnum.DOdiesis), "Nota statica DO# non corrisponde a quella del costruttore")
        Assert.AreEqual(Nota.RE, New Nota(NotaEnum.RE), "Nota statica RE non corrisponde a quella del costruttore")
        Assert.AreEqual(Nota.REdiesis, New Nota(NotaEnum.REdiesis), "Nota statica RE# non corrisponde a quella del costruttore")
        Assert.AreEqual(Nota.MI, New Nota(NotaEnum.MI), "Nota statica MI non corrisponde a quella del costruttore")
        Assert.AreEqual(Nota.FA, New Nota(NotaEnum.FA), "Nota statica FA non corrisponde a quella del costruttore")
        Assert.AreEqual(Nota.FAdiesis, New Nota(NotaEnum.FAdiesis), "Nota statica FA# non corrisponde a quella del costruttore")
        Assert.AreEqual(Nota.SOL, New Nota(NotaEnum.SOL), "Nota statica SOL non corrisponde a quella del costruttore")
        Assert.AreEqual(Nota.SOLdiesis, New Nota(NotaEnum.SOLdiesis), "Nota statica SOL# non corrisponde a quella del costruttore")
        Assert.AreEqual(Nota.LA, New Nota(NotaEnum.LA), "Nota statica LA non corrisponde a quella del costruttore")
        Assert.AreEqual(Nota.LAdiesis, New Nota(NotaEnum.LAdiesis), "Nota statica LA# non corrisponde a quella del costruttore")
        Assert.AreEqual(Nota.SI, New Nota(NotaEnum.SI), "Nota statica SI non corrisponde a quella del costruttore")
    End Sub

    <TestMethod()>
    Public Sub VerificaNoteStatiche_Bemolle()
        Assert.AreEqual(Nota.REb, Nota.DOdiesis, "Nota statica REb non corrisponde a quella del DO#")
        Assert.AreEqual(Nota.MIb, Nota.REdiesis, "Nota statica MIb non corrisponde a quella del RE#")
        Assert.AreEqual(Nota.SOLb, Nota.FAdiesis, "Nota statica SOLb non corrisponde a quella del FA#")
        Assert.AreEqual(Nota.LAb, Nota.SOLdiesis, "Nota statica LAb non corrisponde a quella del SOL#")
        Assert.AreEqual(Nota.SIb, Nota.LAdiesis, "Nota statica SIb non corrisponde a quella del LA#")
    End Sub

    <TestMethod()>
    Public Sub NoteStringParseTest()
        Dim n As New Nota(NotaEnum.LA)
        Assert.IsTrue(Nota.TryParse("DO", n), "Conversione di ""DO"" non riuscita")
        Assert.AreEqual(n, Nota.DO, "Conversione di ""DO"" non corrisponde alla nota DO")
        Assert.IsTrue(Nota.TryParse("DO#", n), "Conversione di ""DO#"" non riuscita")
        Assert.AreEqual(n, Nota.DOdiesis, "Conversione di ""DO#"" non corrisponde alla nota DO#")
        Assert.IsTrue(Nota.TryParse("REb", n), "Conversione di ""REb"" non riuscita")
        Assert.AreEqual(n, Nota.DOdiesis, "Conversione di ""REb"" non corrisponde alla nota DO#")
    End Sub

    <TestMethod()>
    Public Sub VerificaToString_Note()
        Assert.AreEqual("DO", Nota.DO.ToString, True, "ToString della nota DO non corrisponde")
        Assert.AreEqual("DO#", Nota.DOdiesis.ToString, True, "ToString della nota DO# non corrisponde")
        Assert.AreEqual("RE", Nota.RE.ToString, True, "ToString della nota RE non corrisponde")
        Assert.AreEqual("RE#", Nota.REdiesis.ToString, True, "ToString della nota RE# non corrisponde")
        Assert.AreEqual("MI", Nota.MI.ToString, True, "ToString della nota MI non corrisponde")
        Assert.AreEqual("FA", Nota.FA.ToString, True, "ToString della nota FA non corrisponde")
        Assert.AreEqual("FA#", Nota.FAdiesis.ToString, True, "ToString della nota FA# non corrisponde")
        Assert.AreEqual("SOL", Nota.SOL.ToString, True, "ToString della nota SOL non corrisponde")
        Assert.AreEqual("SOL#", Nota.SOLdiesis.ToString, True, "ToString della nota SOL# non corrisponde")
        Assert.AreEqual("LA", Nota.LA.ToString, True, "ToString della nota LA non corrisponde")
        Assert.AreEqual("LA#", Nota.LAdiesis.ToString, True, "ToString della nota LA# non corrisponde")
        Assert.AreEqual("SI", Nota.SI.ToString, True, "ToString della nota SI non corrisponde")
    End Sub

    <TestMethod()>
    Public Sub VerificaToString_NoteBemolle()
        Assert.AreEqual("REb", Nota.DOdiesis.ToString(True), True, "ToString della nota DO# non corrisponde al REb")
        Assert.AreEqual("MIb", Nota.REdiesis.ToString(True), True, "ToString della nota RE# non corrisponde al MIb")
        Assert.AreEqual("SOLb", Nota.FAdiesis.ToString(True), True, "ToString della nota FA# non corrisponde al SOLb")
        Assert.AreEqual("LAb", Nota.SOLdiesis.ToString(True), True, "ToString della nota SOL# non corrisponde al LAb")
        Assert.AreEqual("SIb", Nota.LAdiesis.ToString(True), True, "ToString della nota LA# non corrisponde al SIb")
    End Sub

    <TestMethod()>
    Public Sub VerificaSaltiNote_dalDO()
        Dim C = Nota.DO
        Dim nuova = C + Semitono.Value
        Assert.AreEqual(Nota.DOdiesis, nuova, "Verifica DO + 1 semitono")
        nuova = C + 2 * Semitono.Value
        Assert.AreEqual(Nota.RE, nuova, "Verifica DO + 2 semitoni")
        nuova = C + 3 * Semitono.Value
        Assert.AreEqual(Nota.REdiesis, nuova, "Verifica DO + 3 semitoni")
        For i = NotaEnum.DO To NotaEnum.SI
            nuova = New Nota(i)
            Assert.AreEqual(nuova, C + CInt(i) / 2, String.Format("Verifica costruzione con salti nota DO + {0}", CInt(i)))
        Next
    End Sub

    <TestMethod()>
    Public Sub VerificaSaltiNote_dalDOdiesis()
        Dim C_ = Nota.DOdiesis
        Dim nuova = C_ + Semitono.Value
        Assert.AreEqual(Nota.[RE], nuova, "Verifica DO# + 1 semitono")
        nuova = C_ + 2 * Semitono.Value
        Assert.AreEqual(Nota.REdiesis, nuova, "Verifica DO# + 2 semitoni")
        nuova = C_ + 3 * Semitono.Value
        Assert.AreEqual(Nota.MI, nuova, "Verifica DO# + 3 semitoni")
        For i = NotaEnum.[DO] To NotaEnum.SI
            nuova = New Nota((1 + i) Mod 12)
            Assert.AreEqual(nuova, C_ + CInt(i) / 2, String.Format("Verifica costruzione con salti nota DO# + {0}", CInt(i)))
        Next
    End Sub

    <TestMethod()>
    Public Sub VerificaSaltiNote_dalRE()
        Dim D = Nota.[RE]
        Dim nuova = D + Semitono.Value
        Assert.AreEqual(Nota.REdiesis, nuova, "Verifica RE + 1 semitono")
        nuova = D + 2 * Semitono.Value
        Assert.AreEqual(Nota.MI, nuova, "Verifica RE + 2 semitoni")
        nuova = D + 3 * Semitono.Value
        Assert.AreEqual(Nota.FA, nuova, "Verifica RE + 3 semitoni")
        For i = NotaEnum.[DO] To NotaEnum.SI
            nuova = New Nota((D.Valore + i) Mod 12)
            Assert.AreEqual(nuova, D + CInt(i) / 2, String.Format("Verifica costruzione con salti nota RE + {0}", CInt(i)))
        Next
    End Sub

    <TestMethod()>
    Public Sub VerificaSaltiNote_dalREdiesis()
        Dim D_ = Nota.REdiesis
        Dim nuova = D_ + Semitono.Value
        Assert.AreEqual(Nota.MI, nuova, "Verifica RE# + 1 semitono")
        nuova = D_ + 2 * Semitono.Value
        Assert.AreEqual(Nota.FA, nuova, "Verifica RE# + 2 semitoni")
        nuova = D_ + 3 * Semitono.Value
        Assert.AreEqual(Nota.FAdiesis, nuova, "Verifica RE# + 3 semitoni")
        For i = NotaEnum.[DO] To NotaEnum.SI
            nuova = New Nota((D_.Valore + i) Mod 12)
            Assert.AreEqual(nuova, D_ + CInt(i) / 2, String.Format("Verifica costruzione con salti nota RE + {0}", CInt(i)))
        Next
    End Sub

    <TestMethod()> Public Sub VerificaSaltiNote_dalMI()
        Dim E = Nota.MI
        Dim nuova As Nota
        For i = NotaEnum.[DO] To NotaEnum.SI
            nuova = New Nota((E.Valore + i) Mod 12)
            Assert.AreEqual(nuova, E + CInt(i) / 2, String.Format("Verifica costruzione con salti nota MI + {0}", CInt(i)))
        Next
    End Sub
    <TestMethod()> Public Sub VerificaSaltiNote_dalFA()
        Dim F = Nota.FA
        Dim nuova As Nota
        For i = NotaEnum.[DO] To NotaEnum.SI
            nuova = New Nota((F.Valore + i) Mod 12)
            Assert.AreEqual(nuova, F + CInt(i) / 2, String.Format("Verifica costruzione con salti nota FA + {0}", CInt(i)))
        Next
    End Sub
    <TestMethod()> Public Sub VerificaSaltiNote_dalFAdiesis()
        Dim F_ = Nota.FAdiesis
        Dim nuova As Nota
        For i = NotaEnum.[DO] To NotaEnum.SI
            nuova = New Nota((F_.Valore + i) Mod 12)
            Assert.AreEqual(nuova, F_ + CInt(i) / 2, String.Format("Verifica costruzione con salti nota FA# + {0}", CInt(i)))
        Next
    End Sub
    <TestMethod()> Public Sub VerificaSaltiNote_dalSOL()
        Dim G = Nota.SOL
        Dim nuova As Nota
        For i = NotaEnum.[DO] To NotaEnum.SI
            nuova = New Nota((G.Valore + i) Mod 12)
            Assert.AreEqual(nuova, G + CInt(i) / 2, String.Format("Verifica costruzione con salti nota SOL + {0}", CInt(i)))
        Next
    End Sub
    <TestMethod()> Public Sub VerificaSaltiNote_dalSOLdiesis()
        Dim G_ = Nota.SOLdiesis
        Dim nuova As Nota
        For i = NotaEnum.[DO] To NotaEnum.SI
            nuova = New Nota((G_.Valore + i) Mod 12)
            Assert.AreEqual(nuova, G_ + CInt(i) / 2, String.Format("Verifica costruzione con salti nota SOL# + {0}", CInt(i)))
        Next
    End Sub
    <TestMethod()> Public Sub VerificaSaltiNote_dalLA()
        Dim A = Nota.LA
        Dim nuova As Nota
        For i = NotaEnum.[DO] To NotaEnum.SI
            nuova = New Nota((A.Valore + i) Mod 12)
            Assert.AreEqual(nuova, A + CInt(i) / 2, String.Format("Verifica costruzione con salti nota LA + {0}", CInt(i)))
        Next
    End Sub
    <TestMethod()> Public Sub VerificaSaltiNote_dalLAdiesis()
        Dim A_ = Nota.LAdiesis
        Dim nuova As Nota
        For i = NotaEnum.[DO] To NotaEnum.SI
            nuova = New Nota((A_.Valore + i) Mod 12)
            Assert.AreEqual(nuova, A_ + CInt(i) / 2, String.Format("Verifica costruzione con salti nota LA# + {0}", CInt(i)))
        Next
    End Sub
    <TestMethod()> Public Sub VerificaSaltiNote_dalSI()
        Dim B = Nota.SI
        Dim nuova As Nota
        For i = NotaEnum.[DO] To NotaEnum.SI
            nuova = New Nota((B.Valore + i) Mod 12)
            Assert.AreEqual(nuova, B + CInt(i) / 2, String.Format("Verifica costruzione con salti nota SI + {0}", CInt(i)))
        Next
    End Sub


End Class

File: Enzo.Music.Kantaro.Test/UtNota.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using static System.Net.Mime.MediaTypeNames;

namespace Enzo.Music.Kantaro.Test;

internal class UtNota
{
    [Test]
    public void VerificaNoteStatiche()
    {
        Assert.That(new Nota(NotaEnum.DO), Is.EqualTo(Nota.DO), "Nota statica DO non corrisponde a quella del costruttore");
        Assert.That(new Nota(NotaEnum.DOdiesis), Is.EqualTo(Nota.DOdiesis), "Nota statica DO# non corrisponde a quella del costruttore");
        Assert.That(new Nota(NotaEnum.RE), Is.EqualTo(Nota.RE), "Nota statica RE non corrisponde a quella del costruttore");
        Assert.That(new Nota(NotaEnum.REdiesis), Is.EqualTo(Nota.REdiesis), "Nota statica RE# non corrisponde a quella del costruttore");
        Assert.That(new Nota(NotaEnum.MI), Is.EqualTo(Nota.MI), "Nota statica MI non corrisponde a quella del costruttore");
        Assert.That(new Nota(NotaEnum.FA), Is.EqualTo(Nota.FA), "Nota statica FA non corrisponde a quella del costruttore");
        Assert.That(new Nota(NotaEnum.FAdiesis), Is.EqualTo(Nota.FAdiesis), "Nota statica FA# non corrisponde a quella del costruttore");
        Assert.That(new Nota(NotaEnum.SOL), Is.EqualTo(Nota.SOL), "Nota statica SOL non corrisponde a quella del costruttore");
        Assert.That(new Nota(NotaEnum.SOLdiesis), Is.EqualTo(Nota.SOLdiesis), "Nota statica SOL# non corrisponde a quella del costruttore");
        Assert.That(new Nota(NotaEnum.LA), Is.EqualTo(Nota.LA), "Nota statica LA non corrisponde a quella del costruttore");
        Assert.That(new Nota(NotaEnum.LAdiesis), Is.EqualTo(Nota.LAdiesis), "Nota statica LA# non corrisponde a quella del costruttore");
        Assert.That(new Nota(NotaEnum.SI), Is.EqualTo(Nota.SI), "Nota statica SI non corrisponde a quella del costruttore");
    }

    [Test]
    public void VerificaNoteStatiche_Bemolle()
    {
        Assert.That(Nota.DOdiesis, Is.EqualTo(Nota.REb), "Nota statica REb non corrisponde a quella del DO#");
        Assert.That(Nota.REdiesis, Is.EqualTo(Nota.MIb), "Nota statica MIb non corrisponde a quella del RE#");
        Assert.That(Nota.FAdiesis, Is.EqualTo(Nota.SOLb), "Nota statica SOLb non corrisponde a quella del FA#");
        Assert.That(Nota.SOLdiesis, Is.EqualTo(Nota.LAb), "Nota statica LAb non corrisponde a quella del SOL#");
        Assert.That(Nota.LAdiesis, Is.EqualTo(Nota.SIb), "Nota statica SIb non corrisponde a quella del LA#");
    }

    [Test]
    public void NoteStringParseTest()
    {
        Assert.That(Nota.TryParse("DO", out Nota? n), "Conversione di \"DO\" non riuscita");
        Assert.That(Nota.DO, Is.EqualTo(n), "Conversione di \"DO\" non corrisponde alla nota DO");
        Assert.That(Nota.TryParse("DO#", out n), "Conversione di \"DO#\" non riuscita");
        Assert.That(Nota.DOdiesis, Is.EqualTo(n), "Conversione di \"DO#\" non corrisponde alla nota DO#");
        Assert.That(Nota.TryParse("REb", out n), "Conversione di \"REb\" non riuscita");
        Assert.That(Nota.DOdiesis, Is.EqualTo(n), "Conversione di \"REb\" non corrisponde alla nota DO#");
    }

    [Test]
    public void VerificaToString_Note()
    {
        Assert.That(Nota.DO.ToString().ToUpper(), Is.EqualTo("DO"), "ToString della nota DO non corrisponde");
        Assert.That(Nota.DOdiesis.ToString().ToUpper(), Is.EqualTo("DO#"), "ToString della nota DO# non corrisponde");
        Assert.That(Nota.RE.ToString().ToUpper(), Is.EqualTo("RE"), "ToString della nota RE non corrisponde");
        Assert.That(Nota.REdiesis.ToString().ToUpper(), Is.EqualTo("RE#"), "ToString della nota RE# non corrisponde");
        Assert.That(Nota.MI.ToString().ToUpper(), Is.EqualTo("MI"), "ToString della nota MI non corrisponde");
        Assert.That(Nota.FA.ToString().ToUpper(), Is.EqualTo("FA"), "ToString della nota FA non corrisponde");
        Assert.That(Nota.FAdiesis.ToString().ToUpper(), Is.EqualTo("FA#"), "ToString della nota FA# non corrisponde");
        Assert.That(Nota.SOL.ToString().ToUpper(), Is.EqualTo("SOL"), "ToString della nota SOL non corrisponde");
        Assert.That(Nota.SOLdiesis.ToString().ToUpper(), Is.EqualTo("SOL#"), "ToString della nota SOL# non corrisponde");
        Assert.That(Nota.LA.ToString().ToUpper(), Is.EqualTo("LA"), "ToString della nota LA non corrisponde");
        Assert.That(Nota.LAdiesis.ToString().ToUpper(), Is.EqualTo("LA#"), "ToString della nota LA# non corrisponde");
        Assert.That(Nota.SI.ToString().ToUpper(), Is.EqualTo("SI"), "ToString della nota SI non corrisponde");
    }

    [Test]
    public void VerificaToString_NoteBemolle()
    {
        Assert.IsTrue("REb".Equals(Nota.DOdiesis.ToString(true), StringComparison.InvariantCultureIgnoreCase), "ToString della nota DO# non corrisponde al REb");
        Assert.IsTrue("MIb".Equals(Nota.REdiesis.ToString(true), StringComparison.InvariantCultureIgnoreCase), "ToString della nota RE# non corrisponde al MIb");
        Assert.IsTrue("SOLb".Equals(Nota.FAdiesis.ToString(true), StringComparison.InvariantCultureIgnoreCase), "ToString della nota FA# non corrisponde al SOLb");
        Assert.IsTrue("LAb".Equals(Nota.SOLdiesis.ToString(true), StringComparison.InvariantCultureIgnoreCase), "ToString della nota SOL# non corrisponde al LAb");
        Assert.IsTrue("SIb".Equals(Nota.LAdiesis.ToString(true), StringComparison.InvariantCultureIgnoreCase), "ToString della nota LA# non corrisponde al SIb");
    }

    [Test]
    public void VerificaSaltiNote_dalDO()
    {
        var C = Nota.DO;
        var nuova = C + Semitono.Value;
        Assert.That(nuova, Is.EqualTo(Nota.DOdiesis), "Verifica DO + 1 semitono");
        nuova = C + 2 * Semitono.Value;
        Assert.That(nuova, Is.EqualTo(Nota.RE), "Verifica DO + 2 semitoni");
        nuova = C + 3 * Semitono.Value;
        Assert.That(nuova, Is.EqualTo(Nota.REdiesis), "Verifica DO + 3 semitoni");
        for (NotaEnum i = NotaEnum.DO; i <= NotaEnum.SI; i++)
        {
            nuova = new Nota(i);
            Assert.That(C + (Distanza)((float)i / 2), Is.EqualTo(nuova), $"Verifica costruzione con salti nota DO + {(int)i}");
        }
    }

    [Test]
    public void VerificaSaltiNote_dalDOdiesis()
    {
        var C_ = Nota.DOdiesis;
        var nuova = C_ + Semitono.Value;
        Assert.That(nuova, Is.EqualTo(Nota.RE), "Verifica DO# + 1 semitono");
        nuova = C_ + 2 * Semitono.Value;
        Assert.That(nuova, Is.EqualTo(Nota.REdiesis), "Verifica DO# + 2 semitoni");
        nuova = C_ + 3 * Semitono.Value;
        Assert.That(nuova, Is.EqualTo(Nota.MI), "Verifica DO# + 3 semitoni");
        for (NotaEnum i = NotaEnum.DO; i <= NotaEnum.SI; i++)
        {
            nuova = new Nota((NotaEnum)((1 + (int)i) % 12));
            Assert.That(C_ + (Distanza)((float)i / 2), Is.EqualTo(nuova), $"Verifica costruzione con salti nota DO# + {(int)i}");
        }
    }

    [Test]
    public void VerificaSaltiNote_dalRE()
    {
        var D = Nota.RE;
        var nuova = D + Semitono.Value;
        Assert.That(nuova, Is.EqualTo(Nota.REdiesis), "Verifica RE + 1 semitono");
        nuova = D + 2 * Semitono.Value;
        Assert.That(nuova, Is.EqualTo(Nota.MI), "Verifica RE + 2 semitoni");
        nuova = D + 3 * Semitono.Value;
        Assert.That(nuova, Is.EqualTo(Nota.FA), "Verifica RE + 3 semitoni");
        for (NotaEnum i = NotaEnum.DO; i <= NotaEnum.SI; i++)
        {
            nuova = new Nota((NotaEnum)(((int)D.Valore + (int)i) % 12));
            Assert.That(D + (Distanza)((float)i / 2), Is.EqualTo(nuova), $"Verifica costruzione con salti nota RE + {(int)i}");
        }
    }

    [Test]
    public void VerificaSaltiNote_dalREdiesis()
    {
        var D_ = Nota.REdiesis;
        var nuova = D_ + Semitono.Value;
        Assert.That(nuova, Is.EqualTo(Nota.MI), "Verifica RE# + 1 semitono");
        nuova = D_ + 2 * Semitono.Value;
        Assert.That(nuova, Is.EqualTo(Nota.FA), "Verifica RE# + 2 semitoni");
        nuova = D_ + 3 * Semitono.Value;
        Assert.That(nuova, Is.EqualTo(Nota.FAdiesis), "Verifica RE# + 3 semitoni");
        for (NotaEnum i = NotaEnum.DO; i <= NotaEnum.SI; i++)
        {
            nuova = new Nota((NotaEnum)(((int)D_.Valore + (int)i) % 12));
            Assert.That(D_ + (Distanza)((float)i / 2), Is.EqualTo(nuova), $"Verifica costruzione con salti nota RE + {(int)i}");
        }
    }

    [Test]
    public void VerificaSaltiNote_dalMI()
    {
        var E = Nota.MI;
        Nota nuova;
        for (NotaEnum i = NotaEnum.DO; i <= NotaEnum.SI; i++)
        {
            nuova = new Nota((NotaEnum)(((int)E.Valore + (int)i) % 12));
            Assert.That(E + (Distanza)((float)i / 2), Is.EqualTo(nuova), $"Verifica costruzione con salti nota MI + {(int)i}");
        }
    }
    [Test]
    public void VerificaSaltiNote_dalFA()
    {
        var F = Nota.FA;
        Nota nuova;
        for (NotaEnum i = NotaEnum.DO; i <= NotaEnum.SI; i++)
        {
            nuova = new Nota((NotaEnum)(((int)F.Valore + (int)i) % 12));
            Assert.That(F + (Distanza)((float)i / 2), Is.EqualTo(nuova), $"Verifica costruzione con salti nota FA + {(int)i}");
        }
    }
    [Test]
    public void VerificaSaltiNote_dalFAdiesis()
    {
        var F_ = Nota.FAdiesis;
        Nota nuova;
        for (NotaEnum i = NotaEnum.DO; i <= NotaEnum.SI; i++)
        {
            nuova = new Nota((NotaEnum)(((int)F_.Valore + (int)i) % 12));
            Assert.That(F_ + (Distanza)((float)i / 2), Is.EqualTo(nuova), $"Verifica costruzione con salti nota FA# + {(int)i}");
        }
    }
    [Test]
    public void VerificaSaltiNote_dalSOL()
    {
        var G = Nota.SOL;
        Nota nuova;
        for (NotaEnum i = NotaEnum.DO; i <= NotaEnum.SI; i++)
        {
            nuova = new Nota((NotaEnum)(((int)G.Valore + (int)i) % 12));
            Assert.That(G + (Distanza)((float)i / 2), Is.EqualTo(nuova), $"Verifica costruzione con salti nota SOL + {(int)i}");
        }
    }
    [Test]
    public void VerificaSaltiNote_dalSOLdiesis()
    {
        var G_ = Nota.SOLdiesis;
        Nota nuova;
        for (NotaEnum i = NotaEnum.DO; i <= NotaEnum.SI; i++)
        {
            nuova = new Nota((NotaEnum)(((int)G_.Valore + (int)i) % 12));
            Assert.That(G_ + (Distanza)((float)i / 2), Is.EqualTo(nuova), $"Verifica costruzione con salti nota SOL# + {(int)i}");
        }
    }
    [Test]
    public void VerificaSaltiNote_dalLA()
    {
        var A = Nota.LA;
        Nota nuova;
        for (NotaEnum i = NotaEnum.DO; i <= NotaEnum.SI; i++)
        {
            nuova = new Nota((NotaEnum)(((int)A.Valore + (int)i) % 12));
            Assert.That(A + (Distanza)((float)i / 2), Is.EqualTo(nuova), $"Verifica costruzione con salti nota LA + {(int)i}");
        }
    }
    [Test]
    public void VerificaSaltiNote_dalLAdiesis()
    {
        var A_ = Nota.LAdiesis;
        Nota nuova;
        for (NotaEnum i = NotaEnum.DO; i <= NotaEnum.SI; i++)
        {
            nuova = new Nota((NotaEnum)(((int)A_.Valore + (int)i) % 12));
            Assert.That(A_ + (Distanza)((float)i / 2), Is.EqualTo(nuova), $"Verifica costruzione con salti nota LA# + {(int)i}");
        }
    }
    [Test]
    public void VerificaSaltiNote_dalSI()
    {
        var B = Nota.SI;
        Nota nuova;
        for (NotaEnum i = NotaEnum.DO; i <= NotaEnum.SI; i++)
        {
            nuova = new Nota((NotaEnum)(((int)B.Valore + (int)i) % 12));
            Assert.That(B + (Distanza)((float)i / 2), Is.EqualTo(nuova), $"Verifica costruzione con salti nota SI + {(int)i}");
        }
    }
}
